### PR TITLE
Fix reference equality checks for float

### LINKED ISSUE #32054: Fix `===` for float
Expected output
```ballerina
io:println(42.0 === 42.0);  // true
io:println(1.0 === 12.0);  // false
io:println(float:NaN === float:NaN);  // true
io:println(-0.0 === 0.0);  // false
```

Note we can use [Java's Double.equals](https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html#equals-java.lang.Object-) for this.

Part of #17977

### PR DESCRIPTION
**Pull request by gabilang**
```
io:println(42.0 === 42.0);  // true
io:println(1.0 === 12.0);  // false
io:println(float:NaN === float:NaN);  // true
io:println(-0.0 === 0.0);  // false
```

Fixes #32054

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 1]
```diff
@@ -406,6 +406,18 @@ public static boolean isEqual(Object lhsValue, Object rhsValue) {
...
 
+    /**
+     * Reference equality check for float values.
+     *
+     * @param lhsValue The value on the left-hand side
+     * @param rhsValue The value of the right-hand side
+     * @return True if values are reference equal, else false.
+     */
+
+    public static boolean checkFloatRefEqual(double lhsValue, double rhsValue) {
```
Comment (manuranga) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:417]: lest use the name `checkFloatExactEqual` instead `checkFloatRefEqual`

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 2]
```diff
@@ -406,6 +406,18 @@ public static boolean isEqual(Object lhsValue, Object rhsValue) {
...
+        return Objects.equals(lhsValue, rhsValue);
```
Comment (manuranga) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:418]: Please try to use more specif method to Doubles.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java [hunk 3]
```diff
@@ -846,8 +848,8 @@ private void generateRefNotEqualIns(BIRNonTerminator.BinaryOp binaryIns) {
         } else if (lhsOpType.tag == TypeTags.BYTE && rhsOpType.tag == TypeTags.BYTE) {
             this.mv.visitJumpInsn(IF_ICMPEQ, label1);
         } else if (lhsOpType.tag == TypeTags.FLOAT && rhsOpType.tag == TypeTags.FLOAT) {
-            this.mv.visitInsn(DCMPL);
-            this.mv.visitJumpInsn(IFEQ, label1);
+            this.mv.visitMethodInsn(INVOKESTATIC, TYPE_CHECKER, CHECK_FLOAT_REF_EQUAL, "(DD)Z", false);
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:851]: This is ok. but it is faster to actually generate the check here.

Comment (manuranga) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java:851]: Current way is ok for this PR.

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java]: Shall we directly generate check here.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 4]
```diff
@@ -406,6 +406,18 @@ public static boolean isEqual(Object lhsValue, Object rhsValue) {
...
+    /**
+     * Reference equality check for float values.
+     *
+     * @param lhsValue The value on the left-hand side
+     * @param rhsValue The value of the right-hand side
+     * @return True if values are reference equal, else false.
+     */
+
+    public static boolean checkFloatExactEqual(double lhsValue, double rhsValue) {
+        return Double.valueOf(lhsValue).equals(Double.valueOf(rhsValue));
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:418]: unnecessary unboxing Double.valueOf(rhsValue)

Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:418]:
```suggestion
       return Double.valueOf(lhsValue).equals(rhsValue);
```

**Review by manuranga — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java (hunk 3 above)
Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java]: Lets add that improvement with Fix for https://github.com/ballerina-platform/ballerina-lang/issues/17977

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**